test/e2e/kubelet_deletes.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// Interval to poll /runningpods on a node
+	pollInterval = 1 * time.Second
+)
+
+// waitTillKubeletDeletesPods waits till pods with UID in targetPodUIDs set
+// are visible on the kubelet's /runningpods endpoint.
+func waitTillKubeletDeletesPods(restClient *client.Client, targetPodUIDs util.StringSet, nodeName string, timeout time.Duration) error {
+	return wait.Poll(pollInterval, timeout, func() (bool, error) {
+		pods, err := GetKubeletPods(restClient, nodeName)
+		if err != nil {
+			return false, err
+		}
+
+		alive := 0
+		for _, pod := range pods.Items {
+			if targetPodUIDs.Has(string(pod.UID)) {
+				alive += 1
+			}
+		}
+		if alive > 0 {
+			Logf("Waiting on kubelet to delete %d pods on node %q", alive, nodeName)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+var _ = Describe("Clean up pods on node", func() {
+	var numNodes int
+	framework := NewFramework("kubelet-delete")
+
+	BeforeEach(func() {
+		nodes, err := framework.Client.Nodes().List(labels.Everything(), fields.Everything())
+		expectNoError(err)
+		numNodes = len(nodes.Items)
+	})
+
+	type DeleteTest struct {
+		podsPerNode int
+		timeout     time.Duration
+	}
+
+	deleteTests := []DeleteTest{
+		{podsPerNode: 10, timeout: 3 * time.Minute},
+	}
+
+	for _, itArg := range deleteTests {
+		name := fmt.Sprintf(
+			"kubelet should be able to delete %d pods per node in %v.", itArg.podsPerNode, itArg.timeout)
+		It(name, func() {
+			var labelSelector labels.Selector
+			nodeNames := util.NewStringSet()
+
+			totalPods := itArg.podsPerNode * numNodes
+
+			By(fmt.Sprintf("Creating a RC of %d pods and wait until all pods of this RC are running", totalPods))
+			rcName := "cleanup" + strconv.Itoa(totalPods) + "-" + string(util.NewUUID())
+
+			Expect(RunRC(RCConfig{
+				Client:    framework.Client,
+				Name:      rcName,
+				Namespace: framework.Namespace.Name,
+				Image:     "gcr.io/google_containers/pause:go",
+				Replicas:  totalPods,
+			})).NotTo(HaveOccurred())
+
+			labelSelector = labels.Set(map[string]string{"name": rcName}).AsSelector()
+			pods, err := framework.Client.Pods(framework.Namespace.Name).List(labelSelector, fields.Everything())
+			expectNoError(err)
+
+			// Record the nodes on which the pods of this RC are scheduled, as
+			// well as UIDs of the pods.
+			desired := util.NewStringSet()
+			for _, pod := range pods.Items {
+				nodeNames.Insert(pod.Spec.NodeName)
+				desired.Insert(string(pod.UID))
+			}
+
+			// Perform a sanity check so that we know all desired pods are
+			// running on the nodes according to kubelet.
+			seen := util.NewStringSet()
+			for _, name := range nodeNames.List() {
```

Comment:
Yes, In theory this is possible, but I think this is not likely to happen in a controlled test. RunRC waits until all pods are running before returns, so for this to happen, some pod needs to fail on its own or gets OOM killed.  

Since we are using "non-existence" of the pods from (/runningPods) to determine successful deletions, the sanity check here ensures "/runningPods" and our checking code works correctly.

What I can do to make this more robust are:
 1. Don't construct a list of nodes with pods running, and simply use all nodes (so the list doesn't become stale).
 2. Poll until all pods are running.

What do you think?